aptos-move/aptos-vm/src/block_executor/mod.rs

```diff
@@ -25,26 +26,35 @@ use aptos_types::{
         signature_verified_transaction::SignatureVerifiedTransaction, TransactionOutput,
         TransactionStatus,
     },
-    write_set::WriteOp,
+    vm_status::DiscardedVMStatus,
+    write_set::{TransactionWrite, WriteOp, WriteSet},
 };
 use aptos_vm_logging::{flush_speculative_logs, init_speculative_logs};
 use aptos_vm_types::output::VMOutput;
-use move_core_types::vm_status::VMStatus;
+use bytes::Bytes;
+use claims::assert_ok;
+use move_core_types::{language_storage::StructTag, vm_status::VMStatus};
 use once_cell::sync::OnceCell;
 use rayon::ThreadPool;
-use std::{collections::HashMap, sync::Arc};
+use std::{
+    collections::{BTreeMap, HashMap},
+    sync::Arc,
+};
 
-// Wrapper to avoid orphan rule
+/// Output type wrapper used by block executor. VM output and finalized groups are stored
+/// first, then transformed into TransactionOutput type that is returned.
 #[derive(Debug)]
 pub struct AptosTransactionOutput {
     vm_output: Mutex<Option<VMOutput>>,
+    finalized_groups: Mutex<Option<Vec<(StateKey, WriteOp, Vec<(StructTag, Arc<WriteOp>)>)>>>,
```

Comment:
Why not have this as part of VMOutput and TransactionOutput?
So, there is one WriteOp for each StructTag, and additionally one WriteOp for the entire StateKey? Why is that?